From c1e1ff6121640d6a07d8c17edd5f15713649182e Mon Sep 17 00:00:00 2001
From: Viktor Ashirov <vashirov@redhat.com>
Date: Sat, 8 Nov 2025 00:59:58 +0100
Subject: [PATCH] ipa-extdom-extop: fix memory leaks

In various code paths the `ctx` structure was freed, but not `ctx`
resources (`base_dn`, `nss_ctx`, `extdom_instance_counter`) which may
have been allocated.

Plugin didn't have SLAPI_PLUGIN_CLOSE_FN registered, so context was
never freed on server shutdown.

Related: https://pagure.io/freeipa/issue/9895
Signed-off-by: Viktor Ashirov <vashirov@redhat.com>
Reviewed-By: Rob Crittenden <rcritten@redhat.com>
Reviewed-By: Alexander Bokovoy <abbra@users.noreply.github.com>
---
 .../ipa-extdom-extop/ipa_extdom_extop.c       | 39 ++++++++++++++++++-
 1 file changed, 38 insertions(+), 1 deletion(-)

diff --git a/daemons/ipa-slapi-plugins/ipa-extdom-extop/ipa_extdom_extop.c b/daemons/ipa-slapi-plugins/ipa-extdom-extop/ipa_extdom_extop.c
index 5d22f9f2d5a107cc265c9d257bdf20918cb74333..a180e3307ce1e9d6b9fb4569c909526f41c01c3d 100644
--- a/daemons/ipa-slapi-plugins/ipa-extdom-extop/ipa_extdom_extop.c
+++ b/daemons/ipa-slapi-plugins/ipa-extdom-extop/ipa_extdom_extop.c
@@ -171,6 +171,26 @@ static int ipa_extdom_start(Slapi_PBlock *pb)
     return LDAP_SUCCESS;
 }
 
+static int ipa_extdom_close(Slapi_PBlock *pb)
+{
+    int ret;
+    struct ipa_extdom_ctx *ctx;
+
+    ret = slapi_pblock_get(pb, SLAPI_PLUGIN_PRIVATE, &ctx);
+    if (ret == 0 && ctx != NULL) {
+        if (ctx->extdom_instance_counter) {
+            slapi_counter_destroy(&ctx->extdom_instance_counter);
+        }
+        if (ctx->nss_ctx) {
+            back_extdom_free_context(&ctx->nss_ctx);
+        }
+        slapi_ch_free_string(&ctx->base_dn);
+        free(ctx);
+    }
+
+    return 0;
+}
+
 static int ipa_extdom_extop(Slapi_PBlock *pb)
 {
     char *oid = NULL;
@@ -360,7 +380,16 @@ static int ipa_extdom_init_ctx(Slapi_PBlock *pb, struct ipa_extdom_ctx **_ctx)
 
 done:
     if (ret) {
-        free(ctx);
+        if (ctx) {
+            if (ctx->extdom_instance_counter) {
+                slapi_counter_destroy(&ctx->extdom_instance_counter);
+            }
+            if (ctx->nss_ctx) {
+                back_extdom_free_context(&ctx->nss_ctx);
+            }
+            slapi_ch_free_string(&ctx->base_dn);
+            free(ctx);
+        }
     } else {
         *_ctx = ctx;
     }
@@ -388,6 +417,10 @@ int ipa_extdom_init(Slapi_PBlock *pb)
         ret = slapi_pblock_set(pb, SLAPI_PLUGIN_START_FN,
                                (void *)ipa_extdom_start);
     }
+    if (!ret) {
+        ret = slapi_pblock_set(pb, SLAPI_PLUGIN_CLOSE_FN,
+                               (void *)ipa_extdom_close);
+    }
     if (!ret) {
         ret = slapi_pblock_set(pb, SLAPI_PLUGIN_EXT_OP_OIDLIST,
                                ipa_extdom_oid_list);
@@ -405,6 +438,10 @@ int ipa_extdom_init(Slapi_PBlock *pb)
     }
     if (ret) {
         LOG("Failed to set plug-in version, function, and OID.\n" );
+        slapi_counter_destroy(&extdom_ctx->extdom_instance_counter);
+        back_extdom_free_context(&extdom_ctx->nss_ctx);
+        slapi_ch_free_string(&extdom_ctx->base_dn);
+        free(extdom_ctx);
         return -1;
     }
 
-- 
2.52.0

